### PR TITLE
New input plugin for Intel RDT (Intel Resource Director Technology)

### PR DESCRIPTION
**Pull request by p-zak**
### Required for all PRs:

- [x] Signed [CCLA](https://www.influxdata.com/legal/ccla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

CCLA completed and signed by Intel Corporation (I hope ICLA is not required since this github account is business account).

This PR solves https://github.com/influxdata/telegraf/issues/8087

## Feature Request

We have created a Telegraf input plugin which collects information provided by monitoring features of Intel Resource Director Technology (Intel<sup>&reg;</sup> RDT) like Cache Monitoring Technology (CMT), Memory Bandwidth Monitoring (MBM), Cache Allocation Technology (CAT) and Code and Data Prioritization (CDP) Technology provide the hardware framework to monitor and control the utilization of shared resources, like last level cache, memory bandwidth. These technologies comprise Intel’s Resource Director Technology (RDT). As multithreaded and multicore platform architectures emerge, running workloads in single-threaded, multithreaded, or complex virtual machine environment, the last level cache and memory bandwidth are key resources to manage. Intel introduces CMT, MBM, CAT and CDP to manage these various workloads across shared resources.

To gather Intel RDT metrics plugin uses _pqos_ cli tool which is a part of [Intel(R) RDT Software Package](https://github.com/intel/intel-cmt-cat).

The plugin we have created will be able to expose following metrics:
- Memory Bandwidth on Local NUMA Node
- Memory Bandwidth on Remote NUMA Node
- Total Memory Bandwidth
- L3 Cache Occupancy
- L3 Cache Misses
- Instructions Per Cycle

### CODE REVIEW

File: plugins/inputs/intel_rdt/intel_rdt.go [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+func (r *IntelRDT) Description() string {
+	return "Intel Resource Director Technology plugin"
+}
+
+func (r *IntelRDT) SampleConfig() string {
+	return `
+	## Optionally set sampling interval to Nx100ms. 
+	## This value is propagated to pqos tool. Interval format is defined by pqos itself.
+	## If not provided or provided 0, will be set to 10 = 10x100ms = 1s.
+	# sampling_interval = "10"
```
Comment (ssoroka) [inline plugins/inputs/intel_rdt/intel_rdt.go:80]: I find this a little unclear. is this the number of samples per second?

Comment (ssoroka) [reply plugins/inputs/intel_rdt/intel_rdt.go:80]: FYI there is also a config.Duration type

File: plugins/inputs/intel_rdt/intel_rdt.go [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+	} else if len(processesPIDsAssociation) != 0 {
+		toOmit = toOmit + len(processesPIDsAssociation)
+	}
+	for omitCounter := 0; omitCounter < toOmit; omitCounter++ {
+		reader.Scan()
+	}
+	for reader.Scan() {
+		out := reader.Text()
+		// to handle situation when monitored PID disappear and "err" is shown in output
+		if strings.Contains(out, "err") {
```
Comment (ssoroka) [inline plugins/inputs/intel_rdt/intel_rdt.go:315]: err seems like a common string segment. any chance it catches non-error text?

**Review by ssoroka — APPROVED** [review]
looks good. Couple comments

File: plugins/inputs/intel_rdt/intel_rdt.go (hunk 1 above)
Comment (KubaTrojan) [reply plugins/inputs/intel_rdt/intel_rdt.go:80]: Thank you for the info about config.Duration. We decided to use sampling_interval to be consistent with pqos tool itself. We provide this configuration to pqos as its mon-interval argument which defines monitoring sampling interval for RDT metrics in 100ms units, where 1=100ms. More info about pqos: https://github.com/intel/intel-cmt-cat/tree/master/pqos

After providing by pqos tool any output data, the plugin parses it immediately and adds to telegraf accumulator. Also, the plugin checks at every interval if desired processes change their PIDs association.

File: plugins/inputs/intel_rdt/intel_rdt.go (hunk 2 above)
Comment (KubaTrojan) [reply plugins/inputs/intel_rdt/intel_rdt.go:315]: Good point. But pqos tool basically returns only numerical-like data in form of CSV. 
There is an example of pqos output:

**Format:**
`Time,PID,Core,IPC,LLC Misses,LLC[KB],MBL[MB/s],MBR[MB/s],MBT[MB/s]`

**Without errorr:**
`2020-08-12 13:34:36,"45417,29170",37,44,0.00,0,0.0,0.0,0.0,0.0`

**With error:**
`2020-08-12 13:34:36,"45417,29170",err,0.00,0,0.0,0.0,0.0,0.0`

The `err` substring in `Core` field is a very specific case showing only when some problem with the monitored process happened.

**Comment by ssoroka** [comment]
Merged! thanks for the contribution :)

**Pull request merged**